refactor(CNV-51887): move libdv out of functional tests

codingben opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does

The libdv package contains NewDataVolume
function options that helps us to construct
a new v1.VirtualMachine.

Currently libdv used only in functional tests
and not importable by other projects, because
tests likely tend to be changed or removed.

Move libdv package to pkg folder that is likely
not to be changed like libvmi package.

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

codingben commented:
/kind cleanup

codingben commented:
/cc @0xFelix

0xFelix commented on tests/clone_test.go:
```diff
@@ -13,8 +13,8 @@ import (
 	"kubevirt.io/kubevirt/pkg/libvmi"
 	"kubevirt.io/kubevirt/pkg/pointer"
 
```
nit: Move to block above

codingben replied:
Done in this [revision](https://github.com/kubevirt/kubevirt/compare/eed4dfd64c3d0d8a8b9fff48933f75ff830f8959..3aa64ad2b9e147718c1376d3bede0ee7eb643671).

0xFelix reviewed (commented):
Important question: Who is going to own libdv? Who is going to maintain it?

If we should agree to move `libdv` to `pkg`, it needs an `OWNER` file and a `README.md`, similar to `libvmi`.

akalenyu reviewed (commented):
> Important question: Who is going to own libdv? Who is going to maintain it?
> 
> If we should agree to move `libdv` to `pkg`, it needs an `OWNER` file and a `README.md`, similar to `libvmi`.

sig-storage could own it, but my worry is that kubevirt/kubevirt as it is today is not really importable.

@xpivarc on one of kubevirt-dev's threads:
>  I would generally discourage usage of any code from Kubevirt. We do not provide any compatibility guruanties (exception being API definitions), even for client-go. We recommend generating the client based on API.

So I am not sure how straightforward it would be to import something from kubevirt/kubevirt.
Maybe you could try importing libvmi and see if that works, before committing to libdv?

codingben commented:
@0xFelix I see that Alex replied to your question about ownership of this pkg.

> Maybe you could try importing libvmi and see if that works, before committing to libdv?

@akalenyu We just did an import of libvmi, and added an additional new function ([this one](https://github.com/kubevirt/kubevirt-tekton-tasks/blob/5df1dedc12a65b33027e99b87cf48677692f329a/test/disk_uploader_test.go#L121-L144)) instead of using libdv's options.

akalenyu commented:
> @0xFelix I see that Alex replied to your question about ownership of this pkg.
> 
> > Maybe you could try importing libvmi and see if that works, before committing to libdv?
> 
> @akalenyu We just did an import of libvmi, and added an additional new function ([this one](https://github.com/kubevirt/kubevirt-tekton-tasks/blob/5df1dedc12a65b33027e99b87cf48677692f329a/test/disk_uploader_test.go#L121-L144)) instead of using libdv's options.

Ok sounds good, in that case, could you add the owners file so sig-storage could "properly" own it?

codingben commented:
> > @0xFelix I see that Alex replied to your question about ownership of this pkg.
> > > Maybe you could try importing libvmi and see if that works, before committing to libdv?
> > 
> > 
> > @akalenyu We just did an import of libvmi, and added an additional new function ([this one](https://github.com/kubevirt/kubevirt-tekton-tasks/blob/5df1dedc12a65b33027e99b87cf48677692f329a/test/disk_uploader_test.go#L121-L144)) instead of using libdv's options.
> 
> Ok sounds good, in that case, could you add the owners file so sig-storage could "properly" own it?

Thanks. Yes, I've added owners file.

0xFelix commented:
I'm still undecided about this. Can we take it to `kubevirt-dev` first to see what others think?

codingben commented:
> I'm still undecided about this. Can we take it to `kubevirt-dev` first to see what others think?

What do you think we can achieve from `kubevirt-dev`? We have @akalenyu's approval from storage team to be the owner of this package. Do you have concerns by placing this package inside kubevirt/pkg at all, or something else?

0xFelix commented:
@akalenyu mentioned this:

> ... my worry is that kubevirt/kubevirt as it is today is not really importable.

Although it works in your case, it is still an issue. I'm not sure if we should promote importing `kubevirt/kubevirt`.

codingben commented:
> @akalenyu mentioned this:
> 
> > ... my worry is that kubevirt/kubevirt as it is today is not really importable.
> 
> Although it works in your case, it is still an issue. I'm not sure if we should promote importing `kubevirt/kubevirt`.

Ack.

codingben commented:
> I'm still undecided about this. Can we take it to `kubevirt-dev` first to see what others think?

Done, @0xFelix @akalenyu please add your opinion here: https://groups.google.com/g/kubevirt-dev/c/FBOrghostS4

codingben commented on pkg/libdv/dv.go:
```diff
@@ -24,16 +24,16 @@ import (
 
 const (
 	dvRandomNameLength = 12
+	cirrosVolumeSize   = "512Mi"
```
I wonder if we should name it as `cirrosVolumeSize` and not `defaultVolumeSize`. What do you think? `StorageWithVolumeSize` can be used to override the volume size if needed.

akalenyu replied:
defaultVolumeSize sounds good

codingben replied:
Done, can you please review it again? :pray:

iholder101 commented on pkg/libdv/dv.go:
```diff
@@ -24,16 +24,16 @@ import (
+	"kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
 
 	"kubevirt.io/kubevirt/pkg/pointer"
-	cd "kubevirt.io/kubevirt/tests/containerdisk"
```
That's great!

iholder101 reviewed (commented):
> > I'm still undecided about this. Can we take it to `kubevirt-dev` first to see what others think?
> 
> Done, @0xFelix @akalenyu please add your opinion here: [groups.google.com/g/kubevirt-dev/c/FBOrghostS4](https://groups.google.com/g/kubevirt-dev/c/FBOrghostS4)

Following the discussion in the mailing list, in which multiple people were in favor of this, and due to the fact that I couldn't find a compelling argument against this, I'll approve the PR.

Thank you @codingben for the PR and @0xFelix for the discussion!

We indeed still need to ensure packages like this (and libvmi) are externally importable, which I'd be happy to see in follow-ups. Either way, after this PR these packages are at least importable by production code and are safe from test-specific code slipping in.

/approve

akalenyu reviewed (commented):
/lgtm
thanks!
